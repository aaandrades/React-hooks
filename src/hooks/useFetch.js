import { useEffect, useRef, useState } from "react";
import { setConstantValue } from "typescript";

export const useFetch = (url) => {
  const initialState = {
    data: null,
    loading: true,
    error: null,
  };

  const isMounted = useRef(true);
  const [data, setData] = useState(initialState);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    // Restart the loading while new service
    setData({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setData({
            ...data,
            loading: false,
            error: null,
            data,
          });
        }
      })
      .catch(() => {
        setConstantValue({
          data: null,
          loading: false,
          error: "Cannot load the info",
        });
      });

    // return () => {
    //   cleanup;
    // };
  }, [url]);

  return data;
};
