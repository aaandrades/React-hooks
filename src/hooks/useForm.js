import {  useState } from "react";

export const useForm = (initialState = {}) => {
  // Set the values
  const [values, setValues] = useState(initialState);

  // Get the value of form and set new Values in input
  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  // when submit form
  const handleSubmit = (e) => {
    e.preventDefault();
  };

  return { values, handleInputChange, handleSubmit };
};
