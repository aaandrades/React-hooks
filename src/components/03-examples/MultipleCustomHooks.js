import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";
import "./examples.css";

export const MultipleCustomHooks = () => {
  const { state: counter, increment } = useCounter(1);
  const request = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { loading, data } = request;
  const { author, quote } = !!data && data[0];

  return (
    <>
      <h3>BreakingBad Quotes</h3>
      {loading ? (
        <div className="alert alert-info text-center">Loading ...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote}</p>
          <footer className="blockquote-footer padding">{author}</footer>
        </blockquote>
      )}
      <button className="btn btn-primary" onClick={ increment }>
        Next Quote
      </button>
    </>
  );
};
