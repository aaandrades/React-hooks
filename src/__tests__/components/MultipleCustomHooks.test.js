import React from "react";
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../hooks/useFetch";

jest.mock("../../hooks/useFetch");

describe("Test of Multiple Custoom Hooks", () => {


  it("should show correctly", () => {
    useFetch.mockReturnValue({ data: null, loading: true, err: null });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should show the info", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Andr√©s Andrade",
          quote: "Hola Mundo",
        },
      ],
      loading: false,
      err: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-0").text().trim()).toBe("Hola Mundo");
  });
});
