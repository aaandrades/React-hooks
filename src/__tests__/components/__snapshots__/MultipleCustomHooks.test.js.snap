// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of Multiple Custoom Hooks should show correctly 1`] = `
<Fragment>
  <h3>
    BreakingBad Quotes
  </h3>
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</Fragment>
`;
